### PR TITLE
[WIP] Normalize Education and Employment models/API endpoints #PLAT-1285

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

Education and employment were up til now stored on the user model, as a list of dictionaries. This PR pulls off that info into their own Education and Employment models that have a FK back to the user model. It also adds new education and employment API endpoints.

## Changes

- Add and populate Education and Employment models
- Add education and employment API endpoints
	- Includes the ability to update ordering using JSONAPI Relationship serializers
- Add new jsonschemas for validating fields
- Update the frontend to use the new API endpoints for updating/changing education and employment
- Lots of tests

## TODO
- Fully removing `jobs` and `schools` usage and references around the code
- Make sure merging users works as expected
- API versioning
	- from the ticket: "For older versions of the API, just include blank objects on the user serialization rather than removing the field entirely."
- Update developer API documentation
- Fix validation that is mistakenly allowing both a start and end time along with ongoing. There's a failing test for this.
- General polishing and sanity checking

## QA Notes

This is mostly backend related, but does update the frontend code to use the new models! This also involves a migration of data so should include some prep that you have some users with multiple education and employment before this migration is run.

Pre-migration:
- Have a user with multiple education and employment entries 
	- Make sure to note their order

Post-migration:
- Make sure that user has all their entries and in the correct order
- Make sure you can add new entries
- Make sure you can drag to re-order your entries on the frontent
- Make sure entries stick around and are in the correct order on page refresh

## Documentation

To be added!

## Side Effects

Shouldn't be many no!

## Ticket

https://openscience.atlassian.net/browse/PLAT-1285

### CODE REVIEW

File: api/base/serializers.py [hunk 1]
```diff
@@ -1788,3 +1789,34 @@ class MaintenanceStateSerializer(ser.ModelSerializer):
...
+        raise NotImplementedError
+
+    def get_absolute_url(self, obj):
+        return obj.absolute_api_v2_url
+
+    def run_validation(self, *args, **kwargs):
+        to_validate = self.initial_data.copy()
+        to_validate.pop('id')
+        to_validate.pop('type')
+        validate_user_json(to_validate, self.schema)
```
Comment (erinspace) [inline api/base/serializers.py]: I struggled with where to add jsonschema validation and settled on here, but there might be a way more obvious place.

File: api/base/serializers.py [hunk 2]
```diff
@@ -1788,3 +1789,34 @@ class MaintenanceStateSerializer(ser.ModelSerializer):
     class Meta:
         model = MaintenanceState
         fields = ('level', 'message', 'start', 'end')
+
+
+class BaseProfileSerializer(JSONAPISerializer):
```
Comment (erinspace) [inline api/base/serializers.py]: The naming of these "base" classes was tricky - ended up using "Profile" for a lot of things that were education and employment related. Feel free to rethink and rename.

File: api/users/serializers.py [hunk 3]
```diff
@@ -328,6 +335,110 @@ class Meta:
...
+    id = ser.CharField(required=False, allow_null=True, source='_id')
+    class Meta:
+        type_ = 'education'
+
+    def get_absolute_url(self, obj):
+        return obj.absolute_api_v2_url
+
+
+class UserEducationRelationshipSerializer(BaseAPISerializer):
+    # TODO - consolodate these into a user relationship serializer?
```
Comment (erinspace) [inline api/users/serializers.py]: There's a lot of repeated code in all of these relationship serializers that could definitely be consolidated

File: api_tests/users/views/user_profile_test_mixin.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        wrong_dates = payload(institution='Yeah!', end_date='2018-01-01', start_date=None)
+        res = app.post_json(list_url, wrong_dates, auth=user.auth, expect_errors=True)
+        assert res.status_code == 400
+
+        # end_date before start_date fails
+        flopped_dates = payload(institution='What', start_date='2019-01-01', end_date='2018-01-01')
+        res = app.post_json(list_url, flopped_dates, auth=user.auth, expect_errors=True)
+        assert res.status_code == 400
+
+        # ongoing with start and end date fails
```
Comment (erinspace) [inline api_tests/users/views/user_profile_test_mixin.py]: This is the currently failing test, that really should be prevented by the `jsonschema` bit:

```
        "ongoing": {
            "oneOf": [{
                    "const": false,
                    "required": ["start_date", "end_date"]
                },
                {
                    "const": true,
                    "required": ["start_date"],
                    "not": {
                        "required": ["end_date"]
                    }
                }
            ],
            "type": "boolean"
        },
```

but it isn't and it's confusing!

**Comment by pattisdr** [comment]
closing - development picking up on this again, will be reticketed.

**Pull request closed without merging**